### PR TITLE
[LLC] Add Dust Limit for bitcoin family

### PR DESCRIPTION
**Pull request by lvndry**
### 📝 Description

More info about what dusting is [here](https://www.gemini.com/cryptopedia/basics-of-crypto-dusting).

- Addition of a dust limit of 546 stats for bitcoin
-  Migration of AmountField from jsx -> tsx

### ❓ Context

- **Impacted projects**: `LLC`
- **Linked resource(s)**: `https://ledgerhq.atlassian.net/browse/LIVE-3373`

### ✅ Checklist

- [ ] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 📸 Demo
Screenshot 2022-10-10 at 12 12 47

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx [hunk 1]
```diff
@@ -1,10 +1,11 @@
-// @flow
+/* eslint-disable @typescript-eslint/no-use-before-define */
```
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx:1]: Is it intended ?

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx [hunk 2]
```diff
@@ -112,11 +111,10 @@ const AmountField = ({
       <RequestAmount
         disabled={!!useAllAmount || walletConnectProxy}
         account={account}
-        validTransactionError={amountError || messageGas}
+        validTransactionError={amountError || messageGas || messageDust}
         validTransactionWarning={amountWarning}
         onChange={onChange}
         value={walletConnectProxy ? transaction.amount : amount}
-        showCountervalue={false}
```
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:119]: Why do we remove this property ?

**Review by hedi-edelbloute — APPROVED** [review]

File: libs/ledger-live-common/src/families/bitcoin/js-getTransactionStatus.ts [hunk 3]
```diff
@@ -109,15 +114,21 @@ const getTransactionStatus = async (
     warnings.feeTooHigh = new FeeTooHigh();
   }
 
-  return Promise.resolve({
+  const crypto = cryptoFactory(a.currency.id as Currency);
```
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/js-getTransactionStatus.ts:117]: I think here it's the definition of 'id' that is wrong as Currency is an enum of all possible currencies

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/crypto/base.ts [hunk 4]
```diff
@@ -136,6 +136,11 @@ class Base implements ICrypto {
     return bs58check.encode(payload);
   }
 
+  // default getDustLimit value - implement this methode in coins to set specific value
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/bitcoin/wallet-btc/crypto/base.ts:139]: typo on method

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/crypto/bitcoin.ts [hunk 5]
```diff
@@ -8,6 +8,8 @@ import { InvalidAddress } from "@ledgerhq/errors";
 import { DerivationModes } from "../types";
 import Base from "./base";
 
+const BTC_DUST_LIMIT = 546; // satoshies
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/bitcoin/wallet-btc/crypto/bitcoin.ts:11]: oh so we decided to set a static value and not take txSize into account ?

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx [hunk 6]
```diff
@@ -14,16 +12,16 @@ import Switch from "~/renderer/components/Switch";
...
-  account: AccountLike,
-  transaction: Transaction,
-  onChangeTransaction: (*) => void,
-  status: TransactionStatus,
-  bridgePending: boolean,
-  t: TFunction,
-  initValue?: BigNumber,
-  walletConnectProxy?: boolean,
-  resetInitValue?: () => void,
+  parentAccount?: Account;
```
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:15]: Is it your prettier that did those changes automatically ?

File: apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx (hunk 1 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx:1]: Yes it is, I can remove the line and update the code to respect the linter. IMO this linter rule is kinda useless

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 2 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:119]: The props is not defined in the children component, did not throw an error before because it was not a tsx file

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 6 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:15]: No I migrated the file to typescript

File: libs/ledger-live-common/src/families/bitcoin/wallet-btc/crypto/bitcoin.ts (hunk 5 above)
Comment (lvndry) [reply libs/ledger-live-common/src/families/bitcoin/wallet-btc/crypto/bitcoin.ts:11]: For now yes as it's the value in most wallets
We can improve it in a 2nd PR though

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 6 above)
Comment (hedi-edelbloute) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:15]: Oh so it was a tsx file using flow as typecheck D:

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx [hunk 7]
```diff
@@ -14,16 +12,16 @@ import Switch from "~/renderer/components/Switch";
...
+  account: AccountLike;
+  transaction: Transaction;
+  onChangeTransaction: (_: any) => void;
```
Comment (lambertkevin) [inline apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:18]: Avoid `any` as well here please 🙏

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx [hunk 8]
```diff
@@ -57,23 +55,24 @@ const AmountField = ({
...
-  let { amount: amountError, gasPrice: messageGas } = errors;
+  let { amount: amountError, dustLimit: messageDust, gasPrice: messageGas } = errors;
   let { amount: amountWarning } = warnings;
 
   // we ignore zero case for displaying field error because field is empty.
 
   if (amount.eq(0) && (bridgePending || !useAllAmount)) {
-    amountError = null;
-    amountWarning = null;
+    amountError = undefined;
```
[the remark this answers is not in the document]
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:74]: Don't think it would change anything here, probably just for typing, **but** (I know it was there before and probably out of scope of this PR) we should probably avoid changing the content of errors returned by the bridge in general and always consider them as constants.

To me it would be worth changing this piece of logic for a memoized constant with conditions inside taking into account this moment where `amount.eq(0)` and feed this/those variable(s) to the `RequestAmount` component.  

TBD if we want this in this PR 🤷‍♂️

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 2 above)
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:119]: Looks like it's not actually a valid prop of `RequestAmount`

**Review by lambertkevin — CHANGES_REQUESTED** [review]

File: apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx [hunk 9]
```diff
@@ -51,7 +52,7 @@ export default function RequestAmount({
   validTransactionError,
   validTransactionWarning,
 }: Props) {
-  const fiatCurrency = useSelector(counterValueCurrencySelector);
+  const fiatCurrency = useSelector<any, any>(counterValueCurrencySelector);
```
[the remark this answers is not in the document]
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx:55]: Agreed, unfortunately we'd have to migrate `~/renderer/reducers/settings` to tsx as well for me to know the type
I've tried but I don't know the codebase enough to make such changes as of now

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 7 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:18]: I don't like using any but everywhere else this function is called the parameters are of type `any`

File: libs/ledger-live-common/src/transaction/common.ts [hunk 10]
```diff
@@ -69,8 +71,11 @@ export const fromTransactionStatusRawCommon = (
 export const toTransactionStatusRawCommon = (
   ts: TransactionStatusCommon
 ): TransactionStatusCommonRaw => ({
-  errors: mapValues(ts.errors, toErrorRaw),
-  warnings: mapValues(ts.warnings, toErrorRaw),
+  errors: mapValues(ts.errors, toErrorRaw) as unknown as Record<string, string>,
```
[the remark this answers is not in the document]
Comment (lvndry) [reply libs/ledger-live-common/src/transaction/common.ts:74]: Not perfect but lodash as several signatures for the method mapValues, typescript can't know as of now which one is used
It's not a good practice to say but not "illegal": https://basarat.gitbook.io/typescript/type-system/type-assertion#double-assertion
Any idea how to avoid this ?

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx]: Agreed but I think it's a better idea to do this in a different PR

File: libs/ledger-live-common/src/transaction/common.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (lambertkevin) [reply libs/ledger-live-common/src/transaction/common.ts:74]: Why are you adding those types in the first place ? Cause on my VSCode all the types are inferred automatically.

Anyway `mapValues` has multiple typescript signatures using templating, so I guess you should probably take a loot at that:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/29d4592066a1587254230035f5fbb7c5af46ef19/types/lodash/common/object.d.ts#L1518

`mapValues<Record<string, string>, Error>(ts.errors, toErrorRaw)` would be the correct type here I think.

File: apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx (hunk 1 above)
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx:1]: You should never remove completely an eslint rule from a file unless it being 100% necessary. 
Either update the code, if not possible use `eslint-disable-next-line` where needed, and only if repeated 10 times then it's worth discussing with the team about globally removing it from our eslint rules. 

In this specific example, it's purely styling related, so let's respect it for the sake of code unity.

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 7 above)
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx]: No it's not, actually you can see here the real type of `onChangeTransaction`: https://github.com/LedgerHQ/ledger-live/blob/7806de7ef3f402e1b87df8ac635c49d00c832fff/apps/ledger-live-desktop/src/renderer/modals/Send/types.js#L27

File: apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (lambertkevin) [reply apps/ledger-live-desktop/src/renderer/components/RequestAmount.tsx:55]: Just remove `OutputSelector<State, void, Currency>` from `~/renderer/reducers/settings` to get the right types

File: apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx (hunk 6 above)
Comment (hedi-edelbloute) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:15]: Ok let's be careful and make sure it works with the generated component mechanic

Comment (hedi-edelbloute) [reply apps/ledger-live-desktop/src/renderer/modals/Send/fields/AmountField.tsx:15]: sync-families-dispatch.mjs

**Review by hedi-edelbloute — APPROVED** [review]

**Review by henrily-ledger — APPROVED** [review]

File: libs/ledger-live-common/.eslintrc.js [hunk 11]
```diff
@@ -52,11 +52,11 @@ module.exports = {
...
-          "{}": false
-        }
-      }
-    ]
+        extendDefaults: true,
+        types: {
+          "{}": false,
+        },
+      },
+    ],
```
Comment (lambertkevin) [inline libs/ledger-live-common/.eslintrc.js:60]: Weird that you had to change that, lint on save maybe ?

**Review by lambertkevin — COMMENTED** [review]
LGTM 👍

**Review by lambertkevin — APPROVED** [review]
LGTM but don't forget your changeset !

File: apps/ledger-live-desktop/static/i18n/en/app.json [hunk 12]
```diff
@@ -4953,6 +4953,9 @@
     "FeeTooHigh": {
       "title": "Network fees are above 10% of the amount"
     },
+    "DustLimit": {
+      "title": "Amount is lower than dust limit"
+    },
```
Comment (haammar-ledger) [inline apps/ledger-live-desktop/static/i18n/en/app.json:4958]: Neat :ok_hand:  you just need to do the same on LLM, [here](https://github.com/LedgerHQ/ledger-live/blob/develop/apps/ledger-live-mobile/src/locales/en/common.json).

File: apps/ledger-live-desktop/src/renderer/reducers/settings.js [hunk 13]
```diff
@@ -381,7 +381,7 @@ export const deepLinkUrlSelector = (state: State) => state.settings.deepLinkUrl;
 export const counterValueCurrencyLocalSelector = (state: SettingsState): Currency =>
   findCurrencyByTicker(state.counterValue) || getFiatCurrencyByTicker("USD");
 
-export const counterValueCurrencySelector: OutputSelector<State, void, Currency> = createSelector(
+export const counterValueCurrencySelector = createSelector(
```
Comment (haammar-ledger) [inline apps/ledger-live-desktop/src/renderer/reducers/settings.js:384]: What's the reason for this change? TS complaining about it?

**Review by haammar-ledger — CHANGES_REQUESTED** [review]
Just need the translation key on LLM, the rest looks all good :ok_hand: 

EDIT : also a changeset as mentioned by Kevin :+1:  You can refer to this wiki page: https://github.com/LedgerHQ/ledger-live/wiki/Changesets

File: apps/ledger-live-desktop/src/renderer/reducers/settings.js (hunk 13 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/reducers/settings.js:384]: Yes TS complaning. I did this changed based on this PR comment: https://github.com/LedgerHQ/ledger-live/pull/1513#discussion_r991531248

File: apps/ledger-live-mobile/src/screens/SendFunds/03a-AmountCoin.tsx [hunk 14]
```diff
@@ -159,7 +160,10 @@ export default function SendAmountCoin({ navigation, route }: Props) {
                 onChange={onChange}
                 value={amount}
                 error={
-                  amount.eq(0) && (bridgePending || !transaction.useAllAmount)
+                  status.errors.dustLimit
+                    ? status.errors.dustLimit
+                    : amount.eq(0) &&
+                      (bridgePending || !transaction.useAllAmount)
                     ? null
                     : status.errors.amount
```
Comment (haammar-ledger) [inline apps/ledger-live-mobile/src/screens/SendFunds/03a-AmountCoin.tsx:168]: Minor optim, but maybe a `const error = useMemo(...)`?

**Review by haammar-ledger — APPROVED** [review]

File: apps/ledger-live-mobile/src/screens/SendFunds/03a-AmountCoin.tsx (hunk 14 above)
Comment (lvndry) [reply apps/ledger-live-mobile/src/screens/SendFunds/03a-AmountCoin.tsx:168]: Mmm I don't think this is the intented usage of `useMemo` as it's mostly usefull to cache heavy calculation at every rerender (e.g when no props changed, API calls). I'm more in favor of not overusing hooks as a solution to core performance problems (too many rerenders for exemple)

Comment (haammar-ledger) [reply apps/ledger-live-mobile/src/screens/SendFunds/03a-AmountCoin.tsx:168]: Some argument about memoizing everything https://attardi.org/why-we-memo-all-the-things/ ^^
But really nonimportant, nbd.
